### DISCUSSION TITLE
Better HTTP status code for MapQueryString (RequestPayloadValueResolver)

### DISCUSSION THREAD

**Discussion by zim32** [symfony/symfony#50766] [Q&A]
Right now RequestPayloadValueResolver returns 404 (Not found) if GET parameters validation has errors.
Maybe it is better to return 400 which more clearly indicate that parameters passed are invalid?

**Comment by MatTheCat** [comment]
See https://github.com/symfony/symfony/pull/49134#issuecomment-1476251293

**Comment by zim32** [comment]
Is it really hard to add some parameter to MapQueryString f.e. errorStatusCode: 404 or 400 or whatever so anybody can choose whatever fit their needs?

Don't get me wrong but framework is just a tool, not someone's personal opinion. If standard doesn't clarify which code must be used, you should give some way to customize this behaviour. 

Because one day search engine specialist will come and say: hey guys wee have 404 on this URLs, we need to make 400 instead. This is business requirement

**Reply by MatTheCat** [reply]
This is open source; you can open an issue and even submit a PR!

Also note that you can use `Request::$query` to implement your own logic.

**Reply by zim32** [reply]
Yeah this is just discussion to validate the idea itself before actually doing something )

**Comment by zim32** [comment]
Also, query parameters not always identify some resource (which can exist 200, or not 404). Sometimes they represent something which have different semantic (valid / not valid)

**Comment by zim32** [comment]
Another possible solution can be just throwing some exception, f.e. RequestValidationException and let people write custom handling using symfony. But this will be a BC i guess

**Comment by darius-v** [comment]
I am just getting 500 error if there is validation error in symfony 7. Did something change? https://symfony.com/blog/new-in-symfony-6-3-mapping-request-data-to-typed-objects?utm_source=Symfony+Blog+Feed&utm_medium=feed&ref=dailydev there it writes it should not be giving 500 error

**Comment by mlcpro** [comment]
Yes same problem here, I did the test with `MapRequestPayload` and `MapQueryString`, with `Assert\NotNull`, `MapRequestPayload` will return a **422** while `MapQueryString` **404**, this is not logical, it should also return a **422**

**Comment by BenMakesGames** [comment]
agreed that 404 is a weird response code for a validation failure.

you can customize the response code using the `validationFailedStatusCode` parameter:

```php
#[MapQueryString(validationFailedStatusCode: Response::HTTP_UNPROCESSABLE_ENTITY)]
```

and if you hate typing all that out every time, you could extend the `MapQueryString` attribute to make your own; a simple example:

```php
#[Attribute(Attribute::TARGET_PARAMETER)]
class MapValidatedQuery extends MapQueryString
{
    public function __construct()
    {
        parent::__construct(validationFailedStatusCode: Response::HTTP_UNPROCESSABLE_ENTITY);
    }
}
```

with ^that, you can use `#[MapValidatedQuery]` instead of `#[MapQueryString]`.

(note that I left out the other three constructor parameters from `MapQueryString`. if you need any/all of those, add them back in.)
